Implemented new Test::* paradigm

lejeunerenard opened the pull request:
I rewrote `Test::Pretty` to use the new `Test::Stream` API in `Test::More`'s alphas. So far all tests pass with Test::More 1.301001_084. Note that the rewrite is not backwards compatible. If it needs to be, let me know and I'll update the PR.

The code has been structured so that all events emitted by a stream are processed via `stream_listener()`. This way tests can be printed recursively (e.g. subtests). `stream_listener()` adds "pretty" output for 3 different events:
1. `subtest` events are printed as "pretty" `ok()`s if subtests are not set to 'delayed'. Otherwise they print their name and then recurse through their child tests. They do not end with an `ok()` anymore.
2. `ok` events are printed same as before.
3. `plan` events are printed if they are a `SKIP`. The rest are ignored.

Everything else is delegated to it's default rendering.

Closes #25 

_Notes:_
- Added `TEST_PRETTY_INDENT` env variable to set indention
- `subtest` names are no longer half indented.
- To differentiate between a "pretty" ok and the dummy tap, I created a new event called `Test::Stream::Event::DummyTap` which outputs the same dummy tap as before. This utilizes the same filehandle as `Test::More` instead of directly writing to `STDOUT`.
- Removed code that set the exit code so that `Test::More` can do that for us.
- Todo items inside a subtest do not know they are in todo given the context is hidden by `Test::Stream`.

lejeunerenard committed:
Basic ok() listener implemented

Cannot figure out how to listen to subtest's events

lejeunerenard committed:
Added substream listener using recursion

lejeunerenard committed:
Reinstate reporting overall success or failure of test

lejeunerenard committed:
Added src_line back in

lejeunerenard committed:
Added diag() back in using $context

lejeunerenard committed:
Added skip rendering for SKIP Blocks

lejeunerenard committed:
Added skip_all rendering

lejeunerenard committed:
Changed from builder to stream in END and returned proper exit code

lejeunerenard committed:
Fixed up stream diag() in END

lejeunerenard committed:
Added catch all for Stream events in Stream listener

lejeunerenard committed:
WIP - Backtracking with follow_up() callback

lejeunerenard committed:
Revert "WIP - Backtracking with follow_up() callback"

This reverts commit 67370b6654a425a2451b64e6a48998e5a0d7f663.

lejeunerenard committed:
Cleaned comments and changed END to follow_up

lejeunerenard committed:
Made exceptions in follow_up for subtests and done_testing

lejeunerenard committed:
Made DummyTap is closer to origin format

lejeunerenard committed:
t/Util.pm appends outputs to temp file

Without this STDERR and STDOUT could clobber each other

lejeunerenard committed:
Removed label from follow_up

lejeunerenard committed:
Removed exit code function and diag for no plan & no done_testing

Test::Stream seems to output the exit code correctly. Test::Stream also
outputs the diag for tests with have no plan and dont call done_testing.

lejeunerenard committed:
Removed src line warning

lejeunerenard committed:
Removed debugging info and unnecessary comments

lejeunerenard committed:
Added subtest_tap => 'delayed' back in

lejeunerenard committed:
Added Todo comments back in

Note that child tests of a subtest will not know that they are in a
todo. This is because Test::Stream hides the fact from them.

tokuhirom commented on lib/Test/Stream/Event/DummyTap.pm:
```diff
@@ -0,0 +1,32 @@
+package Test::Stream::Event::DummyTap;
```
I guess, if you want to use this name space, you need to discus with @exodist first.

exodist replied:
The Test::Stream::Event::\* namespace is intended to be open to additions. I do not foresee any new ones as part of Test-Simple in the near future. If I need one and someone beats me to it I can pick a new name :-)

In other words, +1

exodist commented on lib/Test/Stream/Event/DummyTap.pm:
```diff
@@ -0,0 +1,32 @@
+   ctx_method => 'dummy_tap',
+);
+
+sub init {
```
no need to write your own init

exodist commented on lib/Test/Stream/Event/DummyTap.pm:
```diff
@@ -0,0 +1,32 @@
+   return [ OUT_STD, "\n$ok\n", ];
+}
+
+sub extra_details {
```
No need to write this method, though if you want to check the value in Test::Stream::Tester you need to flesh it out more..

lejeunerenard committed:
Tested DummyTap

lejeunerenard committed:
Use Test::Stream::ExitMagic::Context to determine if done_testing ran

lejeunerenard committed:
Removed $in_subtest as it is no longer required

Before this was required to check if the follow_up had been fired by the
end of a subtest. With this fix,
https://github.com/Test-More/test-more/commit/0850e52eb74e4049bd0cfc36d29b5638b321e73b,
it is no longer necessary.

lejeunerenard committed:
Removed obsolete checks on $called_by_done_testing

zoffixznet commented:
Hey all. I got this dist assigned to me as part of the [CPAN PR Challenge](http://cpan-prc.org/).

This seems to be a large PR. Is there any way to have it merged before I work through the open Issues?

Travis failure on this PR is due to too low version of the prereq specified.

Let me know. Thanks!

lejeunerenard commented:
@zoffixznet I believe @tokuhirom hasn't merged the PR because it relies on an alpha version of the Test::More API. I am not sure when the new API is going to be released, but it makes sense to me to not merge this PR into master until then. The PR breaks backwards compatibility.

If you make any changes to master that can be applied to this PR as well, I'd be willing to help integrate them so it's not lost if this PR is merged.

zoffixznet commented:
Sounds good!

preaction commented:
Is this ready to go? The Test2 module is out, and #36 and #37 are reported failures to install Test::Pretty. Is there anything I can do to help get this out the door?

tokuhirom commented:
hmm the test suite is still failing.

tokuhirom commented:
Travis says following comment. Did you forgot to add the deps?

```
Test/Stream.pm in @INC (@INC contains: /home/travis/build/tokuhirom/Test-Pretty/blib/lib /home/travis/build/tokuhirom/Test-Pretty/blib/arch /home/travis/build/tokuhirom/Test-Pretty/_build/lib /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3 /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3 .) at /home/travis/build/tokuhirom/Test-Pretty/blib/lib/Test/Pretty.pm
```

exodist commented:
You need to use Test2::API not Test::Stream.
On Jun 17, 2016 5:15 PM, "Tokuhiro Matsuno" notifications@github.com
wrote:

> Travis says following comment. Did you forgot to add the deps?
> 
> Test/Stream.pm in @INC (@INC contains: /home/travis/build/tokuhirom/Test-Pretty/blib/lib /home/travis/build/tokuhirom/Test-Pretty/blib/arch /home/travis/build/tokuhirom/Test-Pretty/_build/lib /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/site_perl/5.16.3 /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3/x86_64-linux /home/travis/perl5/perlbrew/perls/5.16/lib/5.16.3 .) at /home/travis/build/tokuhirom/Test-Pretty/blib/lib/Test/Pretty.pm
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> https://github.com/tokuhirom/Test-Pretty/pull/28#issuecomment-226909066,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAEgumKDx-JzKvSuQyUD67GiyqSZwGepks5qMziFgaJpZM4DHho4
> .

preaction commented:
I can take a shot at making that change unless someone with more experience has some time

lejeunerenard commented:
@preaction feel free to tackle this. I haven't touched `Test2` since it was `Test::Stream`.

karupanerura commented:
How is the progress?

tokuhirom commented:
@exodist how do i use Test2::API?

exodist commented:
I think at this point the best thing to do is write a new formatter. Look at https://metacpan.org/source/EXODIST/Test-Simple-1.302067/lib/Test2/Formatter/TAP.pm  as an example implementation.

All you, or anyone else would have to do is load Test2::Formatter::YourFormatter, and it would do the job.

The formatter receives events. Each event will implement the following methods to help you decide what to display (Taken from https://metacpan.org/source/EXODIST/Test-Simple-1.302067/lib/Test2/Event.pm)

```perl
sub causes_fail      { return $bool }       # not ok
sub increments_count { return $bool } # Bumps the test count
sub diagnostics      { return $bool }      # Is a diagnostics message (STDERR)
sub no_display       { return $bool }      # Should not be displayed by most formatters

# Plan details: $max is the count, #directive is something like 'skip all' if set, and 'reason' is the reason to skip all
sub sets_plan { return ($max, $directive, $reason) }
 
# Basically an $event->to_string that a formatter can use to decide what to show, for diag events this is the diagnostics message.
sub summary { return "display this" }
```

So you just write a formatter that has a write($event, $number) method, that prints the right thing to STDERR/STDOUT when the event comes in. Loading the formatter is usually enough to make it be THE formatter.

magnolia-k commented:
rewrote according to Test2::Formatter's API.

https://github.com/tokuhirom/Test-Pretty/pull/38
